### PR TITLE
Replace custom compile-time string creation by BOOST_METAPARSE_STRING

### PR DESCRIPTION
**Pull request by bernhardmgruber**
This PR improves the readability of error messages during compilation by shrinking the template argument list of `pmacc::meta::String` to the necessary amount of characters. It uses `BOOST_METAPARSE_STRING` to convert a string literal into a char pack.

Example error before this change (typo in `pmacc::Frame`):
```
/mnt/c/dev/picongpu/include/picongpu/../pmacc/particles/memory/frames/Frame.hpp:74:47: error: no type named 'asdf' in 'pmacc::ParticleDescription<pmacc::meta::String<'e', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00'>, pmacc::math::CT::Vector<mpl_::integral_c<int, 8>, mpl_::integral_c<int, 8>, mpl_::integral_c<int, 4>>, boost::mpl::v_item<pmacc::localCellIdx, boost::mpl::v_item<picongpu::weighting, boost::mpl::v_item<picongpu::momentum, boost::mpl::v_item<picongpu::position<picongpu::position_pic>, boost::mpl::vector0<>, 0>, 0>, 0>, 0>, boost::mpl::v_item<picongpu::chargeRatio<picongpu::ChargeRatioElectrons>, boost::mpl::v_item<picongpu::massRatio<picongpu::MassRatioElectrons>, boost::mpl::v_item<picongpu::current<picongpu::currentSolver::EmZ<picongpu::particles::shapes::PQS>>, boost::mpl::v_item<picongpu::interpolation<picongpu::FieldToParticleInterpolation<picongpu::particles::shapes::PQS, picongpu::AssignedTrilinearInterpolation>>, boost::mpl::v_item<picongpu::shape<picongpu::particles::shapes::PQS>, boost::mpl::v_item<picongpu::particlePusher<picongpu::particles::pusher::Boris>, boost::mpl::vector0<>, 0>, 0>, 0>, 0>, 0>, 0>, pmacc::HandleGuardRegion<pmacc::particles::policies::ExchangeParticles, picongpu::particles::boundary::CallPluginsAndDeleteParticles>>'
        typedef typename ParticleDescription::asdf Name;
                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~
/mnt/c/dev/picongpu/include/picongpu/../pmacc/particles/memory/buffers/ParticlesBuffer.hpp:132:39: note: in instantiation of template class 'pmacc::Frame<pmacc::ParticlesBuffer<pmacc::ParticleDescription<pmacc::meta::String<'e', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00'>, pmacc::math::CT::Vector<mpl_::integral_c<int, 8>, mpl_::integral_c<int, 8>, mpl_::integral_c<int, 4>>, boost::mpl::v_item<picongpu::weighting, boost::mpl::v_item<picongpu::momentum, boost::mpl::v_item<picongpu::position<picongpu::position_pic>, boost::mpl::vector0<>, 0>, 0>, 0>, boost::mpl::v_item<picongpu::chargeRatio<picongpu::ChargeRatioElectrons>, boost::mpl::v_item<picongpu::massRatio<picongpu::MassRatioElectrons>, boost::mpl::v_item<picongpu::current<picongpu::currentSolver::EmZ<picongpu::particles::shapes::PQS>>, boost::mpl::v_item<picongpu::interpolation<picongpu::FieldToParticleInterpolation<picongpu::particles::shapes::PQS, picongpu::AssignedTrilinearInterpolation>>, boost::mpl::v_item<picongpu::shape<picongpu::particles::shapes::PQS>, boost::mpl::v_item<picongpu::particlePusher<picongpu::particles::pusher::Boris>, boost::mpl::vector0<>, 0>, 0>, 0>, 0>, 0>, 0>, pmacc::HandleGuardRegion<pmacc::particles::policies::ExchangeParticles, picongpu::particles::boundary::CallPluginsAndDeleteParticles>>, pmacc::math::CT::Vector<mpl_::integral_c<int, 8>, mpl_::integral_c<int, 8>, mpl_::integral_c<int, 4>>, picongpu::DeviceHeap, 3>::OperatorCreatePairStaticArray<1>, pmacc::ParticleDescription<pmacc::meta::String<'e', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00'>, pmacc::math::CT::Vector<mpl_::integral_c<int, 8>, mpl_::integral_c<int, 8>, mpl_::integral_c<int, 4>>, boost::mpl::v_item<pmacc::localCellIdx, boost::mpl::v_item<picongpu::weighting, boost::mpl::v_item<picongpu::momentum, boost::mpl::v_item<picongpu::position<picongpu::position_pic>, boost::mpl::vector0<>, 0>, 0>, 0>, 0>, boost::mpl::v_item<picongpu::chargeRatio<picongpu::ChargeRatioElectrons>, boost::mpl::v_item<picongpu::massRatio<picongpu::MassRatioElectrons>, boost::mpl::v_item<picongpu::current<picongpu::currentSolver::EmZ<picongpu::particles::shapes::PQS>>, boost::mpl::v_item<picongpu::interpolation<picongpu::FieldToParticleInterpolation<picongpu::particles::shapes::PQS, picongpu::AssignedTrilinearInterpolation>>, boost::mpl::v_item<picongpu::shape<picongpu::particles::shapes::PQS>, boost::mpl::v_item<picongpu::particlePusher<picongpu::particles::pusher::Boris>, boost::mpl::vector0<>, 0>, 0>, 0>, 0>, 0>, 0>, pmacc::HandleGuardRegion<pmacc::particles::policies::ExchangeParticles, picongpu::particles::boundary::CallPluginsAndDeleteParticles>>>' requested here
            SizeOfOneBorderElement = (sizeof(FrameTypeBorder) + sizeof(BorderFrameIndex))
                                      ^
/mnt/c/dev/picongpu/include/picongpu/../pmacc/particles/ParticlesBase.hpp:61:26: note: in instantiation of template class 'pmacc::ParticlesBuffer<pmacc::ParticleDescription<pmacc::meta::String<'e', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00'>, pmacc::math::CT::Vector<mpl_::integral_c<int, 8>, mpl_::integral_c<int, 8>, mpl_::integral_c<int, 4>>, boost::mpl::v_item<picongpu::weighting, boost::mpl::v_item<picongpu::momentum, boost::mpl::v_item<picongpu::position<picongpu::position_pic>, boost::mpl::vector0<>, 0>, 0>, 0>, boost::mpl::v_item<picongpu::chargeRatio<picongpu::ChargeRatioElectrons>, boost::mpl::v_item<picongpu::massRatio<picongpu::MassRatioElectrons>, boost::mpl::v_item<picongpu::current<picongpu::currentSolver::EmZ<picongpu::particles::shapes::PQS>>, boost::mpl::v_item<picongpu::interpolation<picongpu::FieldToParticleInterpolation<picongpu::particles::shapes::PQS, picongpu::AssignedTrilinearInterpolation>>, boost::mpl::v_item<picongpu::shape<picongpu::particles::shapes::PQS>, boost::mpl::v_item<picongpu::particlePusher<picongpu::particles::pusher::Boris>, boost::mpl::vector0<>, 0>, 0>, 0>, 0>, 0>, 0>, pmacc::HandleGuardRegion<pmacc::particles::policies::ExchangeParticles, picongpu::particles::boundary::CallPluginsAndDeleteParticles>>, pmacc::math::CT::Vector<mpl_::integral_c<int, 8>, mpl_::integral_c<int, 8>, mpl_::integral_c<int, 4>>, picongpu::DeviceHeap, 3>' requested here
        typedef typename BufferType::FrameType FrameType;
```
After this change:
```
/mnt/c/dev/picongpu/include/picongpu/../pmacc/particles/memory/frames/Frame.hpp:74:47: error: no type named 'asdf' in 'pmacc::ParticleDescription<pmacc::meta::String<'e'>, pmacc::math::CT::Vector<mpl_::integral_c<int, 8>, mpl_::integral_c<int, 8>, mpl_::integral_c<int, 4>>, boost::mpl::v_item<pmacc::localCellIdx, boost::mpl::v_item<picongpu::weighting, boost::mpl::v_item<picongpu::momentum, boost::mpl::v_item<picongpu::position<picongpu::position_pic>, boost::mpl::vector0<>, 0>, 0>, 0>, 0>, boost::mpl::v_item<picongpu::chargeRatio<picongpu::ChargeRatioElectrons>, boost::mpl::v_item<picongpu::massRatio<picongpu::MassRatioElectrons>, boost::mpl::v_item<picongpu::current<picongpu::currentSolver::EmZ<picongpu::particles::shapes::PQS>>, boost::mpl::v_item<picongpu::interpolation<picongpu::FieldToParticleInterpolation<picongpu::particles::shapes::PQS, picongpu::AssignedTrilinearInterpolation>>, boost::mpl::v_item<picongpu::shape<picongpu::particles::shapes::PQS>, boost::mpl::v_item<picongpu::particlePusher<picongpu::particles::pusher::Boris>, boost::mpl::vector0<>, 0>, 0>, 0>, 0>, 0>, 0>, pmacc::HandleGuardRegion<pmacc::particles::policies::ExchangeParticles, picongpu::particles::boundary::CallPluginsAndDeleteParticles>>'
        typedef typename ParticleDescription::asdf Name;
                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~
/mnt/c/dev/picongpu/include/picongpu/../pmacc/particles/memory/buffers/ParticlesBuffer.hpp:132:39: note: in instantiation of template class 'pmacc::Frame<pmacc::ParticlesBuffer<pmacc::ParticleDescription<pmacc::meta::String<'e'>, pmacc::math::CT::Vector<mpl_::integral_c<int, 8>, mpl_::integral_c<int, 8>, mpl_::integral_c<int, 4>>, boost::mpl::v_item<picongpu::weighting, boost::mpl::v_item<picongpu::momentum, boost::mpl::v_item<picongpu::position<picongpu::position_pic>, boost::mpl::vector0<>, 0>, 0>, 0>, boost::mpl::v_item<picongpu::chargeRatio<picongpu::ChargeRatioElectrons>, boost::mpl::v_item<picongpu::massRatio<picongpu::MassRatioElectrons>, boost::mpl::v_item<picongpu::current<picongpu::currentSolver::EmZ<picongpu::particles::shapes::PQS>>, boost::mpl::v_item<picongpu::interpolation<picongpu::FieldToParticleInterpolation<picongpu::particles::shapes::PQS, picongpu::AssignedTrilinearInterpolation>>, boost::mpl::v_item<picongpu::shape<picongpu::particles::shapes::PQS>, boost::mpl::v_item<picongpu::particlePusher<picongpu::particles::pusher::Boris>, boost::mpl::vector0<>, 0>, 0>, 0>, 0>, 0>, 0>, pmacc::HandleGuardRegion<pmacc::particles::policies::ExchangeParticles, picongpu::particles::boundary::CallPluginsAndDeleteParticles>>, pmacc::math::CT::Vector<mpl_::integral_c<int, 8>, mpl_::integral_c<int, 8>, mpl_::integral_c<int, 4>>, picongpu::DeviceHeap, 3>::OperatorCreatePairStaticArray<1>, pmacc::ParticleDescription<pmacc::meta::String<'e'>, pmacc::math::CT::Vector<mpl_::integral_c<int, 8>, mpl_::integral_c<int, 8>, mpl_::integral_c<int, 4>>, boost::mpl::v_item<pmacc::localCellIdx, boost::mpl::v_item<picongpu::weighting, boost::mpl::v_item<picongpu::momentum, boost::mpl::v_item<picongpu::position<picongpu::position_pic>, boost::mpl::vector0<>, 0>, 0>, 0>, 0>, boost::mpl::v_item<picongpu::chargeRatio<picongpu::ChargeRatioElectrons>, boost::mpl::v_item<picongpu::massRatio<picongpu::MassRatioElectrons>, boost::mpl::v_item<picongpu::current<picongpu::currentSolver::EmZ<picongpu::particles::shapes::PQS>>, boost::mpl::v_item<picongpu::interpolation<picongpu::FieldToParticleInterpolation<picongpu::particles::shapes::PQS, picongpu::AssignedTrilinearInterpolation>>, boost::mpl::v_item<picongpu::shape<picongpu::particles::shapes::PQS>, boost::mpl::v_item<picongpu::particlePusher<picongpu::particles::pusher::Boris>, boost::mpl::vector0<>, 0>, 0>, 0>, 0>, 0>, 0>, pmacc::HandleGuardRegion<pmacc::particles::policies::ExchangeParticles, picongpu::particles::boundary::CallPluginsAndDeleteParticles>>>' requested here
            SizeOfOneBorderElement = (sizeof(FrameTypeBorder) + sizeof(BorderFrameIndex))
                                      ^
/mnt/c/dev/picongpu/include/picongpu/../pmacc/particles/ParticlesBase.hpp:61:26: note: in instantiation of template class 'pmacc::ParticlesBuffer<pmacc::ParticleDescription<pmacc::meta::String<'e'>, pmacc::math::CT::Vector<mpl_::integral_c<int, 8>, mpl_::integral_c<int, 8>, mpl_::integral_c<int, 4>>, boost::mpl::v_item<picongpu::weighting, boost::mpl::v_item<picongpu::momentum, boost::mpl::v_item<picongpu::position<picongpu::position_pic>, boost::mpl::vector0<>, 0>, 0>, 0>, boost::mpl::v_item<picongpu::chargeRatio<picongpu::ChargeRatioElectrons>, boost::mpl::v_item<picongpu::massRatio<picongpu::MassRatioElectrons>, boost::mpl::v_item<picongpu::current<picongpu::currentSolver::EmZ<picongpu::particles::shapes::PQS>>, boost::mpl::v_item<picongpu::interpolation<picongpu::FieldToParticleInterpolation<picongpu::particles::shapes::PQS, picongpu::AssignedTrilinearInterpolation>>, boost::mpl::v_item<picongpu::shape<picongpu::particles::shapes::PQS>, boost::mpl::v_item<picongpu::particlePusher<picongpu::particles::pusher::Boris>, boost::mpl::vector0<>, 0>, 0>, 0>, 0>, 0>, 0>, pmacc::HandleGuardRegion<pmacc::particles::policies::ExchangeParticles, picongpu::particles::boundary::CallPluginsAndDeleteParticles>>, pmacc::math::CT::Vector<mpl_::integral_c<int, 8>, mpl_::integral_c<int, 8>, mpl_::integral_c<int, 4>>, picongpu::DeviceHeap, 3>' requested here
        typedef typename BufferType::FrameType FrameType;
```
Notice the reduction of `pmacc::meta::String<'e', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00', '\x00'>` to `pmacc::meta::String<'e'>` in the error messages.

This PR also contains a fix for #3796.

### CODE REVIEW

**Comment by psychocoderHPC** [comment]
@bernhardmgruber To use `boost::metaparse` do we need to update our [install documentation](https://github.com/ComputationalRadiationPhysics/picongpu/blob/dev/INSTALL.rst#boost)?
I do not think so because I assume all is header-only but I like to be safe.

**Comment by sbastrakov** [comment]
I love it! I think those template errors are giving nightmares to pretty much all PIConGPU developers and users

**Comment by sbastrakov** [comment]
@psychocoderHPC i don't see how header-onlyness changes that it is a dependency? We still have to request it in cmake. If a system provides a boost module with this library a well, it will all work automatically. But it's still a dependency.

**Comment by bernhardmgruber** [comment]
> @bernhardmgruber To use `boost::metaparse` do we need to update our [install documentation](https://github.com/ComputationalRadiationPhysics/picongpu/blob/dev/INSTALL.rst#boost)?
> I do not think so because I assume all is header-only but I like to be safe.

`boost::metaparse` is header-only, see [here](https://www.boost.org/doc/libs/1_77_0/more/getting_started/unix-variants.html#header-only-libraries). Therefore, it is automatically available when you link `Boost::boost` in CMake. No change required.

**Comment by psychocoderHPC** [comment]
CI error with clang as CUDA compiler: https://gitlab.com/hzdr/crp/picongpu/-/jobs/1594017128#L2778

**Comment by bernhardmgruber** [comment]
> CI error with clang as CUDA compiler: https://gitlab.com/hzdr/crp/picongpu/-/jobs/1594017128#L2778

Looks like a bug in Boost.Config: https://github.com/boostorg/config/issues/297#issuecomment-921286448

File: include/pmacc/meta/String.hpp [hunk 1]
```diff
@@ -21,27 +21,22 @@
...
 #pragma once
 
-#include <boost/preprocessor/repetition/repeat_from_to.hpp>
+#include <boost/config.hpp>
+#if defined(__clang__) && defined(__CUDACC__) && defined(BOOST_NO_CXX11_VARIADIC_TEMPLATES)
+// Boost.Config wrongly detects the CUDA version with clang as CUDA compiler and disables variadic templates, which are
+// needed by metaparse. See: https://github.com/boostorg/config/issues/297
+#    undef BOOST_NO_CXX11_VARIADIC_TEMPLATES
+#    undef BOOST_NO_VARIADIC_TEMPLATES
+#endif
```
Comment (bernhardmgruber) [inline include/pmacc/meta/String.hpp:30]: That feels a bit hackish and might need a better solution.

Comment (bernhardmgruber) [reply include/pmacc/meta/String.hpp]: Fixed by adding another workaround to `boost_workaround.hpp`.

**Review by psychocoderHPC — APPROVED** [review]

**Pull request merged**